File: CS/Q138607/MyWizardControl.cs
using DevExpress.XtraWizard;
using DevExpress.XtraEditors;
using DevExpress.XtraNavBar;

namespace DXSample {
    public class MyWizardControl :WizardControl {
        internal DevExpress.XtraNavBar.NavBarControl navBarControl1;
        private DevExpress.XtraNavBar.NavBarGroup navBarGroup1;
        internal PanelControl navBarPanel;

        public MyWizardControl() : base() {
            InitializeComponent();            
        }

        private DisabledItemsVisibility disabledItemsVisibility = DisabledItemsVisibility.Disabled;
        public DisabledItemsVisibility DisabledItemsVisibility {
            get { return disabledItemsVisibility; }
            set { 
                disabledItemsVisibility = value;
                foreach (NavBarItemLink link in navBarGroup1.ItemLinks) {
                    link.Item.Enabled = true;
                    link.Visible = true;
                }
                UpdateLinksStatus();
            }
        }

        private void InitializeComponent() {
            this.navBarControl1 = new DevExpress.XtraNavBar.NavBarControl();
            this.navBarGroup1 = new DevExpress.XtraNavBar.NavBarGroup();
            this.navBarPanel = new PanelControl();
            ((System.ComponentModel.ISupportInitialize)(this.navBarControl1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this)).BeginInit();
            this.SuspendLayout();
            // 
            // navBarControl1
            // 
            this.navBarControl1.ContentButtonHint = null;
            this.navBarControl1.Name = "navBarControl1";
            this.navBarControl1.OptionsNavPane.ExpandedWidth = 140;
            this.navBarControl1.Text = "navBarControl1";
            this.navBarControl1.Groups.Add(navBarGroup1);
            this.navBarControl1.AllowSelectedLink = true;
            this.navBarControl1.Dock = System.Windows.Forms.DockStyle.Fill;
            ((MyWizardViewInfo)ViewInfo).UpdateNavBarBounds();
            navBarPanel.Controls.Add(navBarControl1);
            // 
            // navBarGroup1
            // 
            this.navBarGroup1.Caption = "Steps";
            this.navBarGroup1.Expanded = true;
            this.navBarGroup1.Name = "navBarGroup1";
            this.Controls.Add(navBarPanel);
            ((System.ComponentModel.ISupportInitialize)(this.navBarControl1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this)).EndInit();
            this.ResumeLayout(false);

        }

        private void UpdateLinks() {
            foreach (NavBarItem item in navBarControl1.Items)
                item.LinkClicked -= new NavBarLinkEventHandler(Item_LinkClicked);
            navBarGroup1.ItemLinks.Clear();
            navBarControl1.Items.Clear();
            foreach (BaseWizardPage page in Pages) {
                navBarGroup1.AddItem();
                navBarGroup1.ItemLinks[navBarGroup1.ItemLinks.Count - 1].Item.Caption = page.Text;
                navBarGroup1.ItemLinks[navBarGroup1.ItemLinks.Count - 1].Item.LinkClicked += new NavBarLinkEventHandler(Item_LinkClicked);
            }
            UpdateLinksStatus();
        }

        void Item_LinkClicked(object sender, NavBarLinkEventArgs e) {
            SelectedPageIndex = navBarGroup1.ItemLinks.IndexOf(e.Link);
        }

        private void UpdateLinksStatus() {
            if (SelectedPageIndex < 0 || SelectedPageIndex >= navBarGroup1.ItemLinks.Count) return;
            navBarControl1.SelectedLink = navBarGroup1.ItemLinks[SelectedPageIndex];
            for (int i = 0; i < navBarGroup1.ItemLinks.Count; i++)
                if (i <= SelectedPageIndex) {
                    navBarGroup1.ItemLinks[i].Item.Enabled = true;
                    navBarGroup1.ItemLinks[i].Visible = true;
                } else if (DisabledItemsVisibility == DisabledItemsVisibility.Disabled) navBarGroup1.ItemLinks[i].Item.Enabled = false;
                else navBarGroup1.ItemLinks[i].Visible = false;
        }

        protected override WizardViewInfo CreateViewInfo() { return new MyWizardViewInfo(this); }

        protected override void OnPageAdded(BaseWizardPage page) {
            base.OnPageAdded(page);
            UpdateLinks(); 
        }
        protected override void OnPageRemoved(BaseWizardPage page) {
            base.OnPageRemoved(page);
            UpdateLinks();
        }

        protected override void SetSelectedPageCore(BaseWizardPage prevPage, BaseWizardPage value, Direction direction) {
            base.SetSelectedPageCore(prevPage, value, direction);
            UpdateLinksStatus();
        }
    }

    public enum DisabledItemsVisibility {
        Disabled,
        Hidden
    }
}

File: VB/Q138607/MyWizardControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraWizard
Imports DevExpress.XtraEditors
Imports DevExpress.XtraNavBar

Namespace DXSample
	Public Class MyWizardControl
		Inherits WizardControl
		Friend navBarControl1 As DevExpress.XtraNavBar.NavBarControl
		Private navBarGroup1 As DevExpress.XtraNavBar.NavBarGroup
		Friend navBarPanel As PanelControl

		Public Sub New()
			MyBase.New()
			InitializeComponent()
		End Sub

		Private disabledItemsVisibility_Renamed As DisabledItemsVisibility = DisabledItemsVisibility.Disabled
		Public Property DisabledItemsVisibility() As DisabledItemsVisibility
			Get
				Return disabledItemsVisibility_Renamed
			End Get
			Set(ByVal value As DisabledItemsVisibility)
				disabledItemsVisibility_Renamed = value
				For Each link As NavBarItemLink In navBarGroup1.ItemLinks
					link.Item.Enabled = True
					link.Visible = True
				Next link
				UpdateLinksStatus()
			End Set
		End Property

		Private Sub InitializeComponent()
			Me.navBarControl1 = New DevExpress.XtraNavBar.NavBarControl()
			Me.navBarGroup1 = New DevExpress.XtraNavBar.NavBarGroup()
			Me.navBarPanel = New PanelControl()
			CType(Me.navBarControl1, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' navBarControl1
			' 
			Me.navBarControl1.ContentButtonHint = Nothing
			Me.navBarControl1.Name = "navBarControl1"
			Me.navBarControl1.OptionsNavPane.ExpandedWidth = 140
			Me.navBarControl1.Text = "navBarControl1"
			Me.navBarControl1.Groups.Add(navBarGroup1)
			Me.navBarControl1.AllowSelectedLink = True
			Me.navBarControl1.Dock = System.Windows.Forms.DockStyle.Fill
			CType(ViewInfo, MyWizardViewInfo).UpdateNavBarBounds()
			navBarPanel.Controls.Add(navBarControl1)
			' 
			' navBarGroup1
			' 
			Me.navBarGroup1.Caption = "Steps"
			Me.navBarGroup1.Expanded = True
			Me.navBarGroup1.Name = "navBarGroup1"
			Me.Controls.Add(navBarPanel)
			CType(Me.navBarControl1, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub

		Private Sub UpdateLinks()
			For Each item As NavBarItem In navBarControl1.Items
				RemoveHandler item.LinkClicked, AddressOf Item_LinkClicked
			Next item
			navBarGroup1.ItemLinks.Clear()
			navBarControl1.Items.Clear()
			For Each page As BaseWizardPage In Pages
				navBarGroup1.AddItem()
				navBarGroup1.ItemLinks(navBarGroup1.ItemLinks.Count - 1).Item.Caption = page.Text
				AddHandler navBarGroup1.ItemLinks(navBarGroup1.ItemLinks.Count - 1).Item.LinkClicked, AddressOf Item_LinkClicked
			Next page
			UpdateLinksStatus()
		End Sub

		Private Sub Item_LinkClicked(ByVal sender As Object, ByVal e As NavBarLinkEventArgs)
			SelectedPageIndex = navBarGroup1.ItemLinks.IndexOf(e.Link)
		End Sub

		Private Sub UpdateLinksStatus()
			If SelectedPageIndex < 0 OrElse SelectedPageIndex >= navBarGroup1.ItemLinks.Count Then
				Return
			End If
			navBarControl1.SelectedLink = navBarGroup1.ItemLinks(SelectedPageIndex)
			For i As Integer = 0 To navBarGroup1.ItemLinks.Count - 1
				If i <= SelectedPageIndex Then
					navBarGroup1.ItemLinks(i).Item.Enabled = True
					navBarGroup1.ItemLinks(i).Visible = True
				ElseIf DisabledItemsVisibility = DisabledItemsVisibility.Disabled Then
					navBarGroup1.ItemLinks(i).Item.Enabled = False
				Else
					navBarGroup1.ItemLinks(i).Visible = False
				End If
			Next i
		End Sub

		Protected Overrides Function CreateViewInfo() As WizardViewInfo
			Return New MyWizardViewInfo(Me)
		End Function

		Protected Overrides Sub OnPageAdded(ByVal page As BaseWizardPage)
			MyBase.OnPageAdded(page)
			UpdateLinks()
		End Sub
		Protected Overrides Sub OnPageRemoved(ByVal page As BaseWizardPage)
			MyBase.OnPageRemoved(page)
			UpdateLinks()
		End Sub

		Protected Overrides Sub SetSelectedPageCore(ByVal prevPage As BaseWizardPage, ByVal value As BaseWizardPage, ByVal direction As Direction)
			MyBase.SetSelectedPageCore(prevPage, value, direction)
			UpdateLinksStatus()
		End Sub
	End Class

	Public Enum DisabledItemsVisibility
		Disabled
		Hidden
	End Enum
End Namespace